docs/architecture/adr-064-abci-2.0.md

```diff
@@ -0,0 +1,443 @@
+		return abci.ResponseVerifyVoteExtension{Status: REJECT}
+	}
+
+	if err := ValidatePrices(h.state, req, prices); err != nil {
+		log("failed to validate vote extension", "prices", prices, "err", err)
+		return abci.ResponseVerifyVoteExtension{Status: REJECT}
+	}
+
+	// store updated vote extensions at the given height
+	//
+	// NOTE: Vote extensions can be overridden since we can timeout in a round.
+	SetPrices(h.state, req, req.VoteExtension)
+
+	return abci.ResponseVerifyVoteExtension{Status: ACCEPT}
+}
+```
+
+#### Vote Extension Propagation & Verification
+
+As mentioned previously, vote extensions for height `H` are only made available
+to the proposer at height `H+1` during `PrepareProposal`. However, in order to
+make vote extensions useful, all validators should have access to the agreed upon
+vote extensions at height `H` during `H+1`.
+
+Since CometBFT includes all the vote extension signatures in `RequestPrepareProposal`,
+we propose that the proposing validator manually "inject" the vote extensions
+along with their respective signatures via a special transaction, `VoteExtsTx`,
+into the block proposal during `PrepareProposal`. The `VoteExtsTx` will be
+populated with a single `ExtendedCommitInfo` object which is received directly
+from `RequestPrepareProposal`.
+
+For convention, the `VoteExtsTx` transaction should be the first transaction in
+the block proposal, although chains can implement their own preferences. For
+safety purposes, we also propose that the proposer itself verify all the vote
+extension signatures it receives in `RequestPrepareProposal`.
+
+A validator, upon a `RequestProcessProposal`, will receive the injected `VoteExtsTx`
+which includes the vote extensions along with their signatures. If no such transaction
+exists, the validator MUST REJECT the proposal.
+
+When a validator inspects a `VoteExtsTx`, it will evaluate each `SignedVoteExtension`.
+For each signed vote extension, the validator will generate the signed bytes and
+verify the signature. At least 2/3 valid signatures, based on voting power, must
+be received in order for the block proposal to be valid, otherwise the validator
+MUST REJECT the proposal.
+
+In order to have the ability to validate signatures, `BaseApp` must have access
+to the `x/staking` module, since this module stores an index from consensus
+address to public key. However, we will avoid a direct dependency on `x/staking`
+and instead rely on an interface instead. In addition, the Cosmos SDK will expose
+a default signature verification method which applications can use:
+
+```go
+type ValidatorStore interface {
+	GetValidatorByConsAddr(sdk.Context, cryptotypes.Address) (cryptotypes.PubKey, error)
+}
+
+// ValidateVoteExtensions is a function that an application can execute in
+// ProcessProposal to verify vote extension signatures.
+func (app *BaseApp) ValidateVoteExtensions(ctx sdk.Context, currentHeight int64, extCommit abci.ExtendedCommitInfo) error {
```

Comment:
To confirm - the signature verification logic stated here would be available as a helper method in CosmosSDK right? Trying to avoid the app needing to implement the same logic 😆

Reply:
got it, thanks for clarifying. feel free to resolve this comment!